### PR TITLE
Added Security

### PR DESCRIPTION
**Pull request by AlexVulpine**
Added configuration for Security purposes. Implemented User registration and Authentication. Integrated Jwt token and filter

### CODE REVIEW

**Review by RomanYuzviak — APPROVED** [review]
Looks great!

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Why do your CI checks not work? Why do you delete these files? 
image 
image

File: src/main/java/com/foalex/bookstore/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.SecurityFilterChain;
+import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
+
+@RequiredArgsConstructor
+@Configuration
+@EnableWebSecurity
+@EnableMethodSecurity
```
Comment (liliia-ponomarenko) [inline src/main/java/com/foalex/bookstore/config/SecurityConfig.java]: Do you need  both annotations

File: src/main/java/com/foalex/bookstore/dto/book/BookDto.java [hunk 2]
```diff
@@ -1,21 +1,17 @@
...
+@Data
+public class BookDto {
+    private Long id;
+    private String title;
+    private String author;
+    private String isbn;
+    private BigDecimal price;
+    private String description;
+    private String coverImage;
+    private List<Long> categoryIds;
```
Comment (liliia-ponomarenko) [inline src/main/java/com/foalex/bookstore/dto/book/BookDto.java]: You don’t  have categories  yet

File: src/main/java/com/foalex/bookstore/model/Book.java [hunk 3]
```diff
@@ -1,47 +1,47 @@
...
+import java.math.BigDecimal;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Getter
+@Setter
+@NoArgsConstructor
```
Comment (liliia-ponomarenko) [inline src/main/java/com/foalex/bookstore/model/Book.java]:
```suggestion
```

File: src/main/java/com/foalex/bookstore/model/RoleName.java [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package com.foalex.bookstore.model;
+
+public enum RoleName {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/foalex/bookstore/model/RoleName.java]: Can be nested class of Role

File: src/main/java/com/foalex/bookstore/security/CustomUserDetailsService.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+import com.foalex.bookstore.repository.UserRepository;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Component;
+
+@RequiredArgsConstructor
+@Component
```
Comment (liliia-ponomarenko) [inline src/main/java/com/foalex/bookstore/security/CustomUserDetailsService.java]:
```suggestion
@Service
```

File: src/main/java/com/foalex/bookstore/security/jwt/JwtAuthenticationFilter.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.stereotype.Component;
+import org.springframework.util.StringUtils;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+@RequiredArgsConstructor
+@Component
+public class JwtAuthenticationFilter extends OncePerRequestFilter {
+    private static final String AUTHORIZATION_HEADER = "Authorization";
```
Comment (liliia-ponomarenko) [inline src/main/java/com/foalex/bookstore/security/jwt/JwtAuthenticationFilter.java]: Can be replaced by HttpHeader.AUTHORIZATION

File: src/main/java/com/foalex/bookstore/service/BookService.java [hunk 7]
```diff
@@ -1,18 +1,19 @@
...
+    BookDto save(CreateBookRequestDto requestDto);
+
+    List<BookDto> findAll(Pageable pageable);
+
+    BookDto getById(Long id);
+
+    BookDto update(Long id, CreateBookRequestDto requestDto);
+
+    void delete(Long id);
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/foalex/bookstore/service/BookService.java]:
```suggestion
```

File: src/main/java/com/foalex/bookstore/service/impl/UserServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final RoleRepository repository;
+    private final UserMapper userMapper;
+    private final PasswordEncoder passwordEncoder;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto request)
+            throws RegistrationException {
+        if (userRepository.findByEmail(request.email()).isPresent()) {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/foalex/bookstore/service/impl/UserServiceImpl.java]: Should  be replaced by existByEmail repository method

File: src/main/java/com/foalex/bookstore/validation/FieldsValidator.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+        Object fieldValue = new BeanWrapperImpl(value)
+                .getPropertyValue(field);
+        Object fieldMatchValue = new BeanWrapperImpl(value)
+                .getPropertyValue(fieldMatch);
+
+        if (fieldValue != null) {
+            return fieldValue.equals(fieldMatchValue);
+        } else {
+            return fieldMatchValue == null;
+        }
```
Comment (liliia-ponomarenko) [inline src/main/java/com/foalex/bookstore/validation/FieldsValidator.java]: can be replaced  by the ternary operator

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 10]
```diff
@@ -1,3 +1,11 @@
...
+  - include:
+      file: db/changelog/changes/01-create-books-table.yaml
+  - include:
+      file: db/changelog/changes/02-create-users-table.yaml
+  - include:
+      file: db/changelog/changes/03-create-roles-table.yaml
+  - include:
+      file: db/changelog/changes/04-create-users-roles-table.yaml
+  - include:
+      file: db/changelog/changes/05-add-default-roles-to-db.yaml
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/db.changelog-master.yaml]:
```suggestion
      file: db/changelog/changes/05-add-default-roles-to-db.yaml
      
```

File: src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java [hunk 11]
```diff
@@ -1,85 +1,68 @@
...
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+import org.springframework.web.context.request.ServletWebRequest;
 import org.springframework.web.context.request.WebRequest;
-import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
 
-@ControllerAdvice
-public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {
-    private static final Map<Class<? extends RuntimeException>, HttpStatus> HTTP_STATUS_MAP;
+@Slf4j
```
Comment (liliia-ponomarenko) [inline src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java]: About logs, you use them only in this class. Do you really need to provide logs here? If you use it only for development you can remove it now)

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s improve your solution ;)

File: src/main/java/com/foalex/bookstore/config/SecurityConfig.java (hunk 1 above)
Comment (Elena-Bruyako) [reply src/main/java/com/foalex/bookstore/config/SecurityConfig.java]: Not fixed

File: src/main/java/com/foalex/bookstore/dto/book/BookDto.java (hunk 2 above)
Comment (Elena-Bruyako) [reply src/main/java/com/foalex/bookstore/dto/book/BookDto.java]: Not fixed

File: src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java [hunk 12]
```diff
@@ -1,85 +1,68 @@
...
+    protected ErrorResponseWrapper handleAccessDeniedException(AccessDeniedException ex) {
+        return new ErrorResponseWrapper(LocalDateTime.now(), "forbidden", ex.getMessage());
     }
 
-    private String getErrorMessage(ObjectError error) {
-        return (error instanceof FieldError)
-                ? ((FieldError) error).getField() + " " + error.getDefaultMessage()
-                : error.getDefaultMessage();
+    @ExceptionHandler(value = RegistrationException.class)
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
```
Comment (Elena-Bruyako) [inline src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java]: CONFLICT will be better

File: src/main/java/com/foalex/bookstore/model/Book.java (hunk 3 above)
Comment (Elena-Bruyako) [reply src/main/java/com/foalex/bookstore/model/Book.java]: Not fixed

File: src/main/java/com/foalex/bookstore/model/User.java [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    @Column(name = "last_name",nullable = false)
+    private String lastName;
+
+    @Column(name = "shipping_address", nullable = false)
+    private String shippingAddress;
+
+    @Column(name = "is_deleted", nullable = false)
+    private boolean isDeleted = false;
+
+    @ManyToMany
```
Comment (Elena-Bruyako) [inline src/main/java/com/foalex/bookstore/model/User.java:53]:
```
@EqualsAndHashCode.Exclude
@ToString.Exclude
```

File: src/main/java/com/foalex/bookstore/security/CustomUserDetailsService.java (hunk 5 above)
Comment (Elena-Bruyako) [reply src/main/java/com/foalex/bookstore/security/CustomUserDetailsService.java]: Not fixed

File: src/main/java/com/foalex/bookstore/security/CustomUserDetailsService.java [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
```
Comment (Elena-Bruyako) [inline src/main/java/com/foalex/bookstore/security/CustomUserDetailsService.java]:
```suggestion
    public UserDetails loadUserByUsername(String username) {
```

File: src/main/java/com/foalex/bookstore/security/jwt/JwtAuthenticationFilter.java (hunk 6 above)
Comment (Elena-Bruyako) [reply src/main/java/com/foalex/bookstore/security/jwt/JwtAuthenticationFilter.java]: Not fixed

File: src/main/java/com/foalex/bookstore/service/BookService.java (hunk 7 above)
Comment (Elena-Bruyako) [reply src/main/java/com/foalex/bookstore/service/BookService.java]: Not fixed

File: src/main/java/com/foalex/bookstore/service/impl/BookServiceImpl.java [hunk 15]
```diff
@@ -1,52 +1,59 @@
...
+    @Override
+    public BookDto getById(Long id) {
+        return bookRepository.findById(id)
+                .map(bookMapper::toDto)
+                .orElseThrow(() -> new EntityNotFoundException("Can't find book by id: " + id));
+    }
+
+    @Override
+    public BookDto update(Long id, CreateBookRequestDto requestDto) {
+        Book book = bookMapper.toBook(requestDto);
```
Comment (Elena-Bruyako) [inline src/main/java/com/foalex/bookstore/service/impl/BookServiceImpl.java]: wrong logic 
you should first find book bookRepository.findById(id)
call method bookMapper.updateBookFromDto(book, bookFromDb); (create it) 
then
return bookMapper.toDto(bookRepository.save(book));

File: src/main/java/com/foalex/bookstore/service/impl/BookServiceImpl.java [hunk 16]
```diff
@@ -1,52 +1,59 @@
...
+        book.setId(id);
+        return bookMapper.toDto(bookRepository.save(book));
+    }
+
+    @Override
+    public void delete(Long id) {
+        checkIfBookExistsById(id);
```
Comment (Elena-Bruyako) [inline src/main/java/com/foalex/bookstore/service/impl/BookServiceImpl.java]: remove this method

File: src/main/java/com/foalex/bookstore/service/impl/BookServiceImpl.java [hunk 17]
```diff
@@ -1,52 +1,59 @@
...
+        bookRepository.deleteById(id);
+    }
+
+    private void checkIfBookExistsById(Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/foalex/bookstore/service/impl/BookServiceImpl.java]: remove

File: src/main/java/com/foalex/bookstore/service/impl/UserServiceImpl.java (hunk 8 above)
Comment (Elena-Bruyako) [reply src/main/java/com/foalex/bookstore/service/impl/UserServiceImpl.java]: not fixed

File: src/main/java/com/foalex/bookstore/service/impl/UserServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+            throw new RegistrationException("Registration was failed");
+        }
+        User user = userMapper.toUser(request);
+        user.setPassword(passwordEncoder.encode(request.password()));
+        user.setRoles(getDefaultRoles());
+        return userMapper.toDto(userRepository.save(user));
+    }
+
+    private Set<Role> getDefaultRoles() {
+        return new HashSet<>(Collections.singletonList(
```
Comment (Elena-Bruyako) [inline src/main/java/com/foalex/bookstore/service/impl/UserServiceImpl.java]: return Set.of(repository.findByName(RoleName.USER))

File: src/main/java/com/foalex/bookstore/validation/FieldsValidator.java (hunk 9 above)
Comment (Elena-Bruyako) [reply src/main/java/com/foalex/bookstore/validation/FieldsValidator.java]: not fixed 
return Objects.equals(fieldValue, fieldMatchValue);

File: src/main/resources/db/changelog/db.changelog-master.yaml (hunk 10 above)
Comment (Elena-Bruyako) [reply src/main/resources/db/changelog/db.changelog-master.yaml]: not fixed

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Please fix comment from previous review

File: src/main/java/com/foalex/bookstore/config/SecurityConfig.java (hunk 1 above)
Comment (AlexVulpine) [reply src/main/java/com/foalex/bookstore/config/SecurityConfig.java]: I've looked through the java docs. Now I see that @EnableMethodSecurity actually does all the things of @EnableWebSecurity that I need

File: src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java (hunk 11 above)
Comment (AlexVulpine) [reply src/main/java/com/foalex/bookstore/exception/GlobalExceptionHandler.java]: Yeah, I was using to further debug code a while ago. Deleted it now :)

**Review by liliia-ponomarenko — APPROVED** [review]
Great! 😊

**Pull request merged**